dandi/dandiarchive.py

```diff
@@ -22,8 +16,200 @@
 lgr = get_logger()
 
 
+class ParsedDandiURL(ABC, BaseModel):
+    api_url: AnyHttpUrl  # Does not have a trailing slash
+    dandiset_id: str
+    version_id: Optional[str] = None
+
+    @validator("api_url")
+    def _validate_api_url(cls, v: AnyHttpUrl) -> AnyHttpUrl:
+        return parse_obj_as(AnyHttpUrl, v.rstrip("/"))
+
+    def get_client(self) -> DandiAPIClient:
+        return DandiAPIClient(self.api_url)
+
+    def get_dandiset(self, client: DandiAPIClient) -> dict:
+        return client.get_dandiset(self.dandiset_id, self.get_version_id(client))
+
+    def get_version_id(self, client: DandiAPIClient) -> str:
+        if self.version_id is None:
+            r = client.get(f"/dandisets/{self.dandiset_id}/")
+            published_version = r["most_recent_published_version"]
+            if published_version:
+                return published_version["version"]
+            else:
+                return r["draft_version"]["version"]
+        else:
+            return self.version_id
+
+    @abstractmethod
+    def get_assets(
+        self, client: DandiAPIClient, include_metadata: bool = False
+    ) -> Iterator[dict]:
+        ...
+
+    def get_asset_ids(self, client: DandiAPIClient) -> Iterator[str]:
+        for a in self.get_assets(client):
+            yield a["asset_id"]
+
+    @contextmanager
+    def navigate(
+        self, include_metadata: bool = True
+    ) -> Iterator[Tuple[DandiAPIClient, dict, Iterator[dict]]]:
```

Comment:
please add a docstring describing the purpose and  what it would be yielding in those tuples?

Also since it is the only method which doesn't demand `client` - document that it establishes the session for the client while navigating. 
But if that is how all the other methods are expected to be used, why not to make `.client` to be the lazy property and this one start/stopping the session?

Reply:
I chose to make `client` an argument in order to avoid proliferation of clients in cases where the calling code has its own client instance.

Reply:
Docstring added.